File: src/main/GO/leetcode/lists/hot100/LeetCode139.go
package hot100

func wordBreak(s string, wordDict []string) bool {
	if len(wordDict) == 0 {
		return false
	}
	mp := make(map[string]bool)
	for _, w := range wordDict {
		mp[w] = true
	}
	dp := make([]bool, len(s)+1)
	dp[0] = true
	for i := 0; i <= len(s); i++ {
		for j := 0; j < i; j++ {
			if dp[j] && mp[s[j+1:i]] {
				dp[i] = true
				break
			}
		}
	}
	return dp[len(s)]
}

File: src/main/java/leetcode/lists/hot100/LeetCode139.java
package leetcode.lists.hot100;

import java.util.HashSet;
import java.util.List;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/5 11:30
 */
public class LeetCode139 {

    public boolean wordBreak(String s, List<String> wordDict) {
        HashSet<String> set = new HashSet<>(wordDict);
        boolean[] dp = new boolean[s.length() + 1];
        dp[0] = true;
        for (int i = 1; i <= s.length(); i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] && set.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[dp.length - 1];
    }
}
